crates/fluvio/src/producer/dispatcher.rs

```diff
@@ -0,0 +1,305 @@
+    broadcast_status: tokio::sync::broadcast::Sender<BatchStatus>,
+    /// Pool of individual `PartitionProducer`s, which oct independently
+    pool: HashMap<ReplicaKey, PartitionProducer>,
+}
+
+impl PartitionPool {
+    /// Create a new `PartitionPool`
+    pub fn new(
+        spus: Arc<SpuPool>,
+        config: Arc<ProducerConfig>,
+        broadcast_status: tokio::sync::broadcast::Sender<BatchStatus>,
+    ) -> Self {
+        Self {
+            spus,
+            config,
+            pool: Default::default(),
+            broadcast_status,
+        }
+    }
+
+    /// Buffer the given record by sending it to the appropriate `PartitionProducer`.
+    ///
+    /// If a `PartitionProducer` does not exist for this record's `ReplicaKey`, one
+    /// will be created.
+    pub async fn buffer(
+        &mut self,
+        record: AssociatedRecord,
+        respond_to_caller: Sender<ProducerMsg>,
+    ) -> Result<(), FluvioError> {
+        let partition_producer = match self.pool.get_mut(&record.replica_key) {
+            Some(partition_producer) => partition_producer,
+            None => {
+                let partition_producer = PartitionProducer::new(
+                    self.spus.clone(),
+                    self.config.clone(),
+                    record.replica_key.clone(),
+                    self.broadcast_status.clone(),
+                );
+                self.pool
+                    .insert(record.replica_key.clone(), partition_producer);
+                self.pool
+                    .get_mut(&record.replica_key)
+                    .expect("just inserted, element must exist")
+            }
+        };
+
+        partition_producer.buffer(record, respond_to_caller).await?;
+        Ok(())
+    }
+
+    /// Flush the buffers for all partitions.
+    pub async fn flush_all(&mut self) -> Result<(), FluvioError> {
+        let mut flush_futures = Vec::with_capacity(self.pool.len());
+        for (_, partition_producer) in self.pool.iter_mut() {
+            let fut = partition_producer.flush();
+            flush_futures.push(fut);
+        }
+
+        // Flush all concurrently
+        let results = futures_util::future::join_all(flush_futures).await;
+        for result in results {
+            if let Err(e) = result {
+                error!("Internal error broadcasting batch status: {}", e);
+            }
+        }
+        Ok(())
+    }
+}
+
+/// Manages a buffer for a single partition's records.
+pub(crate) struct PartitionProducer {
+    /// Pool for communicating with the SPUs.
+    spus: Arc<SpuPool>,
+    /// Producer behavior configuration.
+    config: Arc<ProducerConfig>,
+    /// This partition's buffered records.
+    buffer: RecordBuffer,
+    /// A channel for sending the status of batches back to caller.
+    broadcast_status: tokio::sync::broadcast::Sender<BatchStatus>,
+    /// The topic and partition this producer operates for.
+    replica_key: ReplicaKey,
```

Comment:
I think you need state machine here to model behavior of each producer.  they can be:
* Buffering (Ready)
* Sending
* Error

Caller can wait if sending

Reply:
Can we defer this? In practice the code already behaves in these states, I don't think that adding something like a state enum would add much value at this point in time.